libsolidity/ast/Types.cpp

```diff
@@ -2140,12 +2140,21 @@ FunctionType const* ContractType::newExpressionType() const
 	return m_constructorType;
 }
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables() const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(DataLocation _location) const
 {
+	VariableDeclaration::Location location = [&]() {
+		switch (_location)
+		{
+			case DataLocation::Storage: return VariableDeclaration::Location::Unspecified; // By default state variables have unspecified (storage) data location
```

Comment:
I think this should be refactored in a separate PR so variables at storage are explicitly assigned `VariableDeclaration::Location::Storage` (note this is a different enum from `DataLocation` which is in `Types.h`).

Reply:
> Why do we even need two enums for this?

`DataLocation` specifies where the data is stored while `VariableDeclaration::Location`  indicates which modifier is on the variable.

> I think this should be refactored in a separate PR so variables at storage are explicitly assigned `VariableDeclaration::Location::Storage`

`Unspecified` does not necessarily mean `Storage`. That's the case only for state variables. And even then it may also represent a constant or an immutable.

I think changing the way it works would just make things confusing in a different way. Then we'd be asking why it doesn't reflect what's on the variable :) The current usage does have some logic behind it. I think this is just another case of things being too vaguely named, omitting important context. Maybe something like `DeclaredLocation` or `LocationModifier` would be clearer.

> Adding an `Unspecified` option to the `DataLocation` enum would be enough;

It would be misleading because this `Unspecified` would have a different meaning than the one in `VariableDeclaration::Location` and it still would not be a 1:1 translation, IMO `nullopt` is a better choice.